File: src/python/commit_files.py
# commit_files.py
# to run: python3 commit_files.py

import os
import re
import subprocess
from datetime import datetime
import json
import hashlib

def calculate_file_hash(file_path):
    sha256_hash = hashlib.sha256()
    with open(file_path, "rb") as f:
        for byte_block in iter(lambda: f.read(4096), b""):
            sha256_hash.update(byte_block)
    return sha256_hash.hexdigest()

def extract_metadata(content, file_path):
    metadata = {
        'author': '',
        'title': os.path.basename(file_path),
        'hashtags': [],
        'file_hash': calculate_file_hash(file_path)
    }

    # Extract author
    author_match = re.search(r'Author:\s*(.+)', content)
    if author_match:
        metadata['author'] = author_match.group(1)

    # Extract title (assuming it's the first line of the file)
    title_match = re.search(r'^(.+)', content)
    if title_match:
        metadata['title'] = title_match.group(1).strip()

    # Extract hashtags
    metadata['hashtags'] = re.findall(r'#\w+', content)

    return metadata

def store_metadata(file_path, metadata):
    metadata_dir = os.path.join(os.path.dirname(file_path), 'metadata')
    os.makedirs(metadata_dir, exist_ok=True)

    metadata_file = os.path.join(metadata_dir, f"{os.path.basename(file_path)}.json")

    with open(metadata_file, 'w', encoding='utf-8') as f:
        json.dump(metadata, f, indent=2)

    return metadata_file

def run_git_command(command):
    process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
    output, error = process.communicate()
    return output.decode('utf-8').strip(), error.decode('utf-8').strip()

def commit_text_files(repo_path="."):
    os.chdir(repo_path)

    # Check if there are any changes
    status_output, _ = run_git_command("git status --porcelain")
    if not status_output:
        print("No changes to commit.")
        return

    # Get all modified and untracked files
    changed_files, _ = run_git_command("git diff --name-only")
    untracked_files, _ = run_git_command("git ls-files --others --exclude-standard")

    all_files = changed_files.split('\n') + untracked_files.split('\n')
    txt_files = [f for f in all_files if f.endswith('.txt')]

    if not txt_files:
        print("No uncommitted .txt files found.")
        return

    # Process each file and store metadata
    metadata_files = []
    for file_path in txt_files:
        full_path = os.path.join(repo_path, file_path)
        try:
            with open(full_path, 'r', encoding='utf-8') as f:
                content = f.read()

            metadata = extract_metadata(content, full_path)
            metadata_file = store_metadata(full_path, metadata)
            metadata_files.append(metadata_file)

            print(f"File: {file_path}")
            print(f"Author: {metadata['author']}")
            print(f"Title: {metadata['title']}")
            print(f"Hashtags: {', '.join(metadata['hashtags'])}")
            print(f"File Hash: {metadata['file_hash']}")
            print()
        except Exception as e:
            print(f"Error processing file {file_path}: {str(e)}")

    # Add all .txt files and metadata files to staging
    files_to_add = txt_files + metadata_files
    for file in files_to_add:
        run_git_command(f"git add {file}")

    # Create commit message
    commit_message = f"Auto-commit {len(txt_files)} text files and metadata on {datetime.now().strftime('%Y-%m-%d %H:%M:%S')} by commit_files.py"

    # Commit the changes
    run_git_command(f'git commit -m "{commit_message}"')

    print(f"Committed {len(txt_files)} text files and their metadata.")
    print("Commit message:", commit_message)

if __name__ == "__main__":
    commit_text_files(repo_path="message")

File: template/node/commit_files.js
// commit_files.js
// to run: node commit_files.js

const fs = require('fs').promises;
const path = require('path');
const crypto = require('crypto');
const { execSync } = require('child_process');

async function calculateFileHash(filePath) {
	const fileBuffer = await fs.readFile(filePath);
	const hashSum = crypto.createHash('sha256');
	hashSum.update(fileBuffer);
	return hashSum.digest('hex');
}

async function extractMetadata(content, filePath) {
	const metadata = {
		author: '',
		title: path.basename(filePath),
		hashtags: [],
		file_hash: await calculateFileHash(filePath)
	};

	// Extract author
	const authorMatch = content.match(/Author:\s*(.+)/);
	if (authorMatch) {
		metadata.author = authorMatch[1];
	}

	// Extract title (assuming it's the first line of the file)
	const titleMatch = content.match(/^(.+)/);
	if (titleMatch) {
		metadata.title = titleMatch[1].trim();
	}

	// Extract hashtags
	metadata.hashtags = content.match(/#\w+/g) || [];

	return metadata;
}

async function storeMetadata(filePath, metadata) {
	const metadataDir = path.join(path.dirname(filePath), 'metadata');
	await fs.mkdir(metadataDir, { recursive: true });

	const metadataFile = path.join(metadataDir, `${path.basename(filePath)}.json`);

	await fs.writeFile(metadataFile, JSON.stringify(metadata, null, 2), 'utf-8');

	return metadataFile;
}

function gitCommand(command) {
	try {
		return execSync(command, { encoding: 'utf-8' }).trim();
	} catch (error) {
		console.error(`Error executing Git command: ${error.message}`);
		return '';
	}
}

async function commitTextFiles(repoPath = '.') {
	const currDir = process.cwd();
	repoPath = path.resolve(repoPath);
	process.chdir(repoPath);

	try {
		// Get modified and untracked files
		const changedFiles = gitCommand('git diff --name-only').split('\n').filter(Boolean);
		const untrackedFiles = gitCommand('git ls-files --others --exclude-standard').split('\n').filter(Boolean);

		const allFiles = [...changedFiles, ...untrackedFiles];
		const txtFiles = allFiles.filter(f => f.endsWith('.txt'));

		if (!txtFiles.length) {
			console.log("No uncommitted .txt files found.");
			return;
		}

		for (const filePath of txtFiles) {
			try {
				const absPath = path.join(repoPath, filePath);
				const content = await fs.readFile(absPath, 'utf-8');
				const metadata = await extractMetadata(content, absPath);
				const metadataFile = await storeMetadata(absPath, metadata);

				console.log(`File: ${filePath}`);
				console.log(`Author: ${metadata.author}`);
				console.log(`Title: ${metadata.title}`);
				console.log(`Hashtags: ${metadata.hashtags.join(', ')}`);
				console.log(`File Hash: ${metadata.file_hash}`);
				console.log();

				gitCommand(`git add '${filePath}' '${path.relative(repoPath, metadataFile)}'`);
			} catch (e) {
				console.error(`Error processing file ${filePath}: ${e.message}`);
			}
		}

		// Create commit
		const timestamp = new Date().toISOString().replace('T', ' ').split('.')[0];
		const commitMessage = `Auto-commit ${txtFiles.length} text files and metadata on ${timestamp} by commit_files.js`;
		gitCommand(`git commit -m '${commitMessage}'`);

		console.log(`Committed ${txtFiles.length} text files and their metadata.`);
		console.log(`Commit message: ${commitMessage}`);
	} finally {
		process.chdir(currDir);
	}
}

if (require.main === module) {
	const repo_path = (process.argv.length >= 3) ? process.argv[2] : "."
	commitTextFiles(repo_path);
}